text/0000-async.md

```diff
@@ -0,0 +1,774 @@
+    Pending,
+}
+```
+
+When a task returns `Poll::Ready`, the executor knows the task has completed and
+can be dropped.
+
+### Waking up
+
+Each task executor provides its own scheduling facilities, and hence needs to
+customize the way task wakeups are handled. As such, there is a
+`std::task::Wake` trait defining wakeup behavior:
+
+```rust
+/// A way of waking up a specific task.
+///
+/// Any task executor must provide a way of signaling that a task it owns
+/// is ready to be `poll`ed again. Executors do so by providing a wakeup handle
+/// type that implements this trait.
+pub trait Wake: Send + Sync {
+    /// Indicates that the associated task is ready to make progress and should
+    /// be `poll`ed.
+    ///
+    /// Executors generally maintain a queue of "ready" tasks; `wake` should place
+    /// the associated task onto this queue.
+    fn wake(&self);
+}
+```
+
+In general async values are not coupled to any particular executor, so we use a trait
+object to handle waking:
+
+```rust
+/// A `Waker` is a handle for waking up a task by notifying its executor that it
+/// is ready to be run.
+///
+/// This handle contains a trait object pointing to an instance of the `UnsafeWake`
+/// trait, allowing notifications to get routed through it.
+///
+/// Implements `Clone`, `Send`, and `Sync`.
+pub struct Waker { ... }
+
+impl Waker {
+    /// Wake up the task associated with this `Waker`.
+    pub fn wake(&self);
+}
+
+// We will see how to handle the no_std case later in the RFC...
+impl<T> From<T> for Waker where Arc<T>: Wake + 'static { ... }
+```
+
+Task execution always happens in the context of a `Waker` that can be used to
+wake the task up; we'll see the full `core::task::Context` structure below.
+
+### Executors
+
+An executor is responsible for polling tasks to completion. We represent this
+with the `core::task::Executor` trait:
+
+```rust
+/// A task executor.
+///
+/// A *task* is a `()`-producing async value that runs at the top level, and will
+/// be `poll`ed until completion. It's also the unit at which wake-up
+/// notifications occur. Executors, such as thread pools, allow tasks to be
+/// spawned and are responsible for putting tasks onto ready queues when
+/// they are woken up, and polling them when they are ready.
+pub trait Executor {
+    /// Spawn the given task, polling it until completion.
+    ///
+    /// # Errors
+    ///
+    /// The executor may be unable to spawn tasks, either because it has
+    /// been shut down or is resource-constrained.
+    fn spawn_obj(&mut self, task: TaskObj) -> Result<(), SpawnObjError>;
+
+    /// Determine whether the executor is able to spawn new tasks.
+    ///
+    /// # Returns
+    ///
+    /// An `Ok` return means the executor is *likely* (but not guaranteed)
+    /// to accept a subsequent spawn attempt. Likewise, an `Err` return
+    /// means that `spawn` is likely, but not guaranteed, to yield an error.
+    fn status(&self) -> Result<(), SpawnErrorKind> {
```

Comment:
The documentation means there's basically no guarantee whether the return value corresponds to the reality, as there are both false positive and negative. That means you can't reliably use `if x.status().is_xxx()` to check anything in either Ok or Err direction. What is the purpose of this function then?

Reply:
Hints are useful for fast paths.

Reply:
Is there a reason that if status returns an Err, spawn isn't guaranteed to return an Err?

Reply:
Not all error statuses are permanent.

Reply:
Then maybe, SpawnErrorKind should have an API to determine whether or not the executor will/may eventually recover from the error.

Reply:
If it's a hint, the naming should clearly indicate that; e.g. `status_hint()`